proposed/2021/TransitiveDependenciesInPMUI.md

```diff
@@ -0,0 +1,95 @@
+
+There will be an additional property on the PackageItemViewModel that will expose the a list of top-level packages in the dependency graph that originally brought in the transitive dependency. For top-level packages, this property will be an empty list. This list will be used to build the tooltip on transitive packages.
+
+### Transitive Packages details pane
+
+The Details pane will be populated for transitive packages in the same way that it is populated for other installed packages. The same mechanism will be reused for transitive packages so the user package metadata, select a version, and click the Install button to install the package which will promote it to a top-level package. There should not need to be many changes to support adding details for transitive packages.
+
+### A/B Experimentation
+
+**The Transitive Dependencies feature will be rolled out to customers as an A/B Experiment.**
+
+This document is not intended to outline the full details of the A/B Experiment but to highlight some key points about releasing the feature as an experiment.
+
+Running an A/B Experiment with the transitive dependencies feature will allow us to take a data-driven approach to deciding if the feature we built sufficently addresses the desired goal or if we need further investment to positively impact the user experience.
+
+It also allows us the ability to turn off the experiment remotely if we find significant regressions in the experience or remotely broaden the audience if the data shows a positive impact.
+
+We will make use of existing engagement metrics to determine the feature's impact on the user's interaction with the PMUI. Namely, we will look closely at NuGet Actions (Package installs, updates, and uninstalls), as well as PMUI Refresh events on each of the tabs. We will consider adding additional metrics specific to transitive dependencies in consultation with the VS Experimentation team.
+
+
+## Considerations
+
+* The transitive packages list should be collapsed by default because the user may not need to see transitive package information every time they view the Installed tab. There is a potential opportunity to delay processing or querying transitive dependency or the top-level packages that installed the transitive dependencies until after the list is expanded. However, doing so would be premature optimization so this will only be considered if it is shown to have a significant impact on performance. Since the assets file is read from disk into memory, and processing will occur on that in-memory representation, I do not anticipate a significant impact, but this consideration can be visited as an option if we see otherwise.
```

Comment:
> doing so would be premature optimization so this will only be considered if it is shown to have a significant impact on performance. 

Loading all transitives up front, regardless of whether the user wants to see them, doesn't sound to me like a premature optimization. I think that should be part of this spec to consider the impact, and whether there's scenarios where we can pre-load, and some where we should not pre-load transitives.

Reply:
I disagree. Currently, transitive dependencies are already being read in up front because of the PackageRecommender so there will be no change there in continuing to pull in transitive dependencies. It is not the responsibility of this feature to change that behavior.

What this feature will add is an algorithm that follows the top-level dependency to the transitive dependency to build a list of top-level packages that are responsible for installing the transitive dependency. That algorithm will work against an in-memory representation of the assets file. We will need to profile how much time that algorithm takes to determine if needs to be optimized before investing in optimization that will affect both PackageRecommender and adding extra code to delay loading.

Reply:
> That algorithm will work against an in-memory representation of the assets file. 

What will the UI look like when there is not an in-memory representation?

Reply:
That's covered in the Considerations section. We commented on it below.